src/pg/decoder.cr

```diff
@@ -471,6 +471,44 @@ module PG
       end
     end
 
+    struct RangeDecoder
+      include Decoder
+
+      def_oids [
+        3904, # int4range
+      ]
+
+      def decode(io, bytesize, oid)
+        # General format is here: https://git.io/JeWQX
+        # Specifics for the binary representation are here: https://git.io/JeWQ1
+
+        flags = io.read_byte # first byte is flags. TODO: what do they do?
+
+        # TODO: How to use the *_bound_len to inform `read_*` to support int8range
+        lower_bound_len = read_i32(io) # => 4 or 8
+        lower_bound = read_i32(io)
+        upper_bound_len = read_i32(io) # => 4 or 8
+        upper_bound = read_i32(io)
+
+        # some debug cruft here. TODO: remove before merge
+        # puts "flags: #{flags.to_s}"
+        # puts "lbound len: #{lower_bound_len}"
+        # puts "lbound: #{lower_bound}"
+        # puts "ubound len: #{upper_bound_len}"
+        # puts "ubound: #{upper_bound}"
+
+        # TODO: add `exclusive` flag? Do we need to, or since the decoder
+        # is already receiving a discrete range type, maybe we don't?
+        # Ref https://www.postgresql.org/docs/10/rangetypes.html#RANGETYPES-DISCRETE
+        Range.new(lower_bound, upper_bound)
+      end
+
+      def type
+        # TODO: do I actually need a new PG-specific type, or can we use real Crystal Ranges?
+        Range
```

Comment:
Should this be a specifically named `Int4Range`, or should the Range decoder emit a Crystal Range? Is the expectation that the end user desire to be super-specific about whether they are passing an Integer, BigInt, or Numeric type?

Reply:
It's been a while since I've looked into ranges in crystal, but I think maybe postgres ranges have more "features" than crystal ranges. 

In some cases, it is probably easy to normalize a postgres range into a crystal one, like how in the tests you have here the inclusive/exclusive adding 1. However in other cases it might not work, like for the float ranges and date ranges allowing positive and negative infinity. We may have to make our own range class for those. Or enhance the crystal stdlib ranges.

Reply:
`Range` can take `nil` as start and/or end value, so maybe this could be used to represent infinity.